src/jit/optcse.cpp

```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
     // We walk the set of CSE candidates and set the bit corresponsing to the CSEindex
     // in the block's bbCseGen bitset
     //
-    for (unsigned cnt = 0; cnt < optCSECandidateCount; cnt++)
+    for (unsigned inx = 0; inx < optCSECandidateCount; inx++)
     {
-        CSEdsc*      dsc      = optCSEtab[cnt];
+        CSEdsc*      dsc      = optCSEtab[inx];
         unsigned     CSEindex = dsc->csdIndex;
         treeStmtLst* lst      = dsc->csdTreeList;
         noway_assert(lst);
 
         while (lst != nullptr)
         {
-            BasicBlock* block = lst->tslBlock;
-            BitVecOps::AddElemD(cseTraits, block->bbCseGen, genCSEnum2bit(CSEindex));
+            BasicBlock* block   = lst->tslBlock;
+            unsigned    cseBit  = genCSEnum2bit(CSEindex) * 2;
+            unsigned    cseBit2 = cseBit + 1;
+
+            // This CSE is generated in 'block', we always set the cseBit
+            // If this block does not contain a call, we also set cseBit2
+            //
+            // If we have a call in this block the below we walk the trees backwards
+            // to find the CSEs that are generated after the last call in the block.
+            //
+            BitVecOps::AddElemD(cseTraits, block->bbCseGen, cseBit);
+            if ((block->bbFlags & BBF_HAS_CALL) == 0)
+            {
+                BitVecOps::AddElemD(cseTraits, block->bbCseGen, cseBit2);
+            }
             lst = lst->tslNext;
         }
     }
 
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+    {
+        // If the block contains a call we may have to update the bbCseGen
+        // as we may generate some CSEs after the last call in the block.
+        //
+        if ((block->bbFlags & BBF_HAS_CALL) != 0)
```

Comment:
It would be better to make this `if (no call) continue`. It avoids additional indentation in the code below and makes it easy to see that the loop only looks at blocks with calls.

Reply:
OK